sonar-batch/src/main/java/org/sonar/batch/bootstrap/TempFolderProvider.java

```diff
@@ -19,33 +19,109 @@
  */
 package org.sonar.batch.bootstrap;
 
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
 import org.apache.commons.io.FileUtils;
+import org.sonar.api.utils.TempFolder;
 import org.apache.commons.lang.StringUtils;
-import org.picocontainer.injectors.ProviderAdapter;
 import org.sonar.api.CoreProperties;
-import org.sonar.api.utils.TempFolder;
 import org.sonar.api.utils.internal.DefaultTempFolder;
 
-import java.io.File;
 import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.concurrent.TimeUnit;
 
-public class TempFolderProvider extends ProviderAdapter {
+public class TempFolderProvider extends LifecycleProviderAdapter {
+  private static final Logger LOG = Loggers.get(TempFolderProvider.class);
+  private static final long CLEAN_MAX_AGE = TimeUnit.DAYS.toMillis(21);
+  static final String TMP_NAME_PREFIX = ".sonartmp_";
 
-  private TempFolder tempFolder;
+  private DefaultTempFolder tempFolder;
 
   public TempFolder provide(BootstrapProperties bootstrapProps) {
     if (tempFolder == null) {
-      String workingDirPath = StringUtils.defaultIfBlank(bootstrapProps.property(CoreProperties.WORKING_DIRECTORY), CoreProperties.WORKING_DIRECTORY_DEFAULT_VALUE);
-      File workingDir = new File(workingDirPath).getAbsoluteFile();
-      File tempDir = new File(workingDir, ".sonartmp");
+
+      String workingPathName = StringUtils.defaultIfBlank(bootstrapProps.property(CoreProperties.GLOBAL_WORKING_DIRECTORY), CoreProperties.GLOBAL_WORKING_DIRECTORY_DEFAULT_VALUE);
+      Path workingPath = Paths.get(workingPathName).normalize();
+
+      if (!workingPath.isAbsolute()) {
+        Path home = findHome(bootstrapProps);
+        workingPath = home.resolve(workingPath).normalize();
+      }
+
+      try {
+        cleanTempFolders(workingPath);
+      } catch (IOException e) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)